### PR TITLE
 Support stream operation on unpacked array (#4714)

### PR DESCRIPTION
**Pull request by fuad1502**
Trying to solve issue #4714 

- Added support from stream operation on unpacked array
- Correctly handle both wider and narrower stream in compared to target/source variable

### CODE REVIEW

File: include/verilated_funcs.h [hunk 1]
```diff
@@ -1578,6 +1578,12 @@ static inline void VL_ASSIGN_DYN_Q(VlQueue<T>& q, int elem_size, int lbits, QDat
     for (int i = 0; i < size; ++i) q.at(i) = (T)((from >> (i * elem_size)) & mask);
 }
 
+template <typename T, std::size_t T_Depth>
+static inline void VL_ASSIGN_UNPACK_Q(VlUnpacked<T, T_Depth>& q, int elem_size, QData from) {
+    const QData mask = VL_MASK_Q(elem_size);
+    for (int i = 0; i < T_Depth; ++i) q[i] = (T)((from >> ((T_Depth - 1 - i) * elem_size)) & mask);
```
Comment (wsnyder) [inline include/verilated_funcs.h:1584]: And similar below

```suggestion
static inline void VL_ASSIGN_UNPACK_Q(VlUnpacked<T, T_Depth>& q, size_t elem_size, QData from) {
    const QData mask = VL_MASK_Q(elem_size);
    for (size_t i = 0; i < T_Depth; ++i) q[i] = (T)((from >> ((T_Depth - 1 - i) * elem_size)) & mask);
```

File: include/verilated_funcs.h [hunk 2]
```diff
@@ -1592,6 +1598,20 @@ static inline QData VL_DYN_TO_Q(const VlQueue<T>& q, int elem_size) {
...
+    return ret;
+}
+
+template <typename T, std::size_t T_Depth>
+static inline QData VL_UNPACK_TO_Q(const VlUnpacked<T, T_Depth>& q, int elem_size) {
+    QData ret = 0;
+    for (int i = 0; i < T_Depth; ++i) ret |= q[T_Depth - 1 - i] << (i * elem_size);
+    return ret;
+}
+
```
Comment (fuad1502) [inline include/verilated_funcs.h:1614]:
```suggestion
template <typename T, std::size_t T_Depth>
static inline IData VL_UNPACK_TO_I(const VlUnpacked<T, T_Depth>& q, size_t elem_size) {
    QData ret = 0;
    for (size_t i = 0; i < T_Depth; ++i) ret |= q[T_Depth - 1 - i] << (i * elem_size);
    return ret;
}

template <typename T, std::size_t T_Depth>
static inline QData VL_UNPACK_TO_Q(const VlUnpacked<T, T_Depth>& q, size_t elem_size) {
    QData ret = 0;
    for (size_t i = 0; i < T_Depth; ++i) ret |= q[T_Depth - 1 - i] << (i * elem_size);
    return ret;
}

```

**Comment by fuad1502** [comment]
Hi @wsnyder , I just added `t_stream_unpack_wider` in `test_regress`, however, I only compare the result with Vivado Simulator. Is it necessary to check it VCS as well? If yes, can you assist me on this? I don't have a VCS simulator. Thanks!

File: src/V3Const.cpp [hunk 3]
```diff
@@ -2168,6 +2168,21 @@ class ConstVisitor final : public VNVisitor {
...
+                                                 "array to a variable of size greater than 64");
+                }
+                srcp = new AstCvtUnpackArrayToPacked{srcp->fileline(), srcp,
+                                                     nodep->lhsp()->dtypep()};
+                // Handling the case where lhs is wider than rhs by inserting zeros. StreamL does
+                // not require this, since the left streaming operator implicitly handles this.
+                uint32_t packedBits = nodep->lhsp()->widthMin();
+                uint32_t unpackBits
+                    = srcDTypep->arrayUnpackedElements() * srcDTypep->subDTypep()->widthMin();
+                uint32_t offset = packedBits > unpackBits ? packedBits - unpackBits : 0;
```
Comment (wsnyder) [inline src/V3Const.cpp:2183]:
```suggestion
                const uint32_t packedBits = nodep->lhsp()->widthMin();
                const uint32_t unpackBits
                    = srcDTypep->arrayUnpackedElements() * srcDTypep->subDTypep()->widthMin();
                const uint32_t offset = packedBits > unpackBits ? packedBits - unpackBits : 0;
```

File: src/V3Const.cpp [hunk 4]
```diff
@@ -2205,10 +2226,23 @@ class ConstVisitor final : public VNVisitor {
...
+                    srcp
+                        = new AstSel{streamp->fileline(), srcp, sWidth - dstBitWidth, dstBitWidth};
+                }
+                srcp = new AstCvtPackedToUnpackArray{nodep->fileline(), srcp, dstp->dtypep()};
+            } else {
+                if (dWidth == 0) {
+                    srcp = new AstCvtPackedToDynArray{nodep->fileline(), srcp, dstp->dtypep()};
+                } else if (sWidth > dWidth) {
+                    srcp = new AstSel{streamp->fileline(), srcp, sWidth - dWidth, dWidth};
+                }
```
Comment (wsnyder) [inline src/V3Const.cpp:2245]: What if sWidth <= dWidth?   Need an ASSERT_OBJ? Or does it work?

File: src/V3Slice.cpp [hunk 5]
```diff
@@ -226,26 +226,30 @@ class SliceVisitor final : public VNVisitor {
             m_assignError = false;
             if (debug() >= 9) nodep->dumpTree("-  Deslice-In: ");
             AstNodeDType* const dtp = nodep->lhsp()->dtypep()->skipRefp();
+            auto stp = nodep->rhsp();
```
Comment (wsnyder) [inline src/V3Slice.cpp:229]:
```suggestion
            AstNode* stp = nodep->rhsp();
```

**Review by wsnyder — COMMENTED** [review]
With these minor fixes suggested is this good to merge?

File: src/V3Const.cpp (hunk 4 above)
Comment (fuad1502) [reply src/V3Const.cpp:2245]: Ah yes, that's the one last thing we haven't handle before this is good to merge. It should throw an error. I am planning to check this in an earlier stage, probably `V3Width`. So, maybe add an assert here, but throw the error in a previous stage?

Comment (wsnyder) [reply src/V3Const.cpp:2245]: Makes sense.

**Comment by fuad1502** [comment]
If everything looks okay, should be ready to merge.

**Pull request merged**